### PR TITLE
something gan compression depend

### PR DESCRIPTION
**Pull request by ceci3**
flops compute support dygraph 
add conv2d_transpose, instance_norm, layer norm
conv_op split to conv op and elementwise_add op
add unitest

add some dygraph special layers and dygraph common block to paddleslim

### CODE REVIEW

File: paddleslim/analysis/flops.py [hunk 1]
```diff
@@ -13,68 +13,125 @@
...
+                c_out, c_in, k_h, k_w = filter_shape
             _, _, h_out, w_out = output_shape
             # c_in is the channel number of filter. It is (input_channel // groups).
             kernel_ops = k_h * k_w * float(c_in)
-            if len(op.inputs("Bias")) > 0:
-                with_bias = 1
-            else:
-                with_bias = 0
-            op_flops = h_out * w_out * c_out * (kernel_ops + with_bias)
+            ### after dygraph model to static program, conv op donnot have Bias attrs
```
Comment (wanghaoshuang) [inline paddleslim/analysis/flops.py]: Do you forget to deal with the static graph?

File: paddleslim/analysis/flops.py [hunk 2]
```diff
@@ -13,68 +13,125 @@
...
-    return _graph_flops(graph, only_conv=only_conv, detail=detail)
+    return _graph_flops(
+        graph, only_conv=only_conv, only_multiply=only_multiply, detail=detail)
 
 
-def _graph_flops(graph, only_conv=True, detail=False):
+def _graph_flops(graph,
+                 only_conv=True,
+                 only_multiply=False,
+                 dygraph=False,
```
Comment (wanghaoshuang) [inline paddleslim/analysis/flops.py]: It seems that `dygraph` is not used in the current function block.

File: paddleslim/analysis/flops.py [hunk 3]
```diff
@@ -13,68 +13,125 @@
...
+                           ] and not only_conv and not only_multiply:
             input_shape = list(op.inputs("X")[0].shape())
             if input_shape[0] == -1:
                 input_shape[0] = 1
-            flops += np.product(input_shape)
+            op_flops = np.product(input_shape)
+            flops += op_flops
+
+        elif op.type() in ['batch_norm', 'instance_norm', 'layer_norm'
+                           ] and not only_conv and only_multiply:
```
Comment (wanghaoshuang) [inline paddleslim/analysis/flops.py]: Is it multiplication in `batch_norm` or `layer_norm`?

File: paddleslim/analysis/flops.py [hunk 4]
```diff
@@ -13,68 +13,125 @@
...
+
+def flops(program, only_conv=True, only_multiply=False, detail=False):
+    """
+    Get FLOPS of target graph.
     Args:
         program(Program): The program used to calculate FLOPS.
         only_conv(bool): Just return number of mul-adds in convolution and FC layer if `only_conv` is true.
                          default: True.
-        detail(bool): Whether to return detail of each convolution layer.
+        only_multiply(bool): Just return number of muliply in the model if `only_multiply` is true.
```
Comment (wanghaoshuang) [inline paddleslim/analysis/flops.py]: Please give a list of operators considered while only_multiply is true.

File: paddleslim/core/layers.py [hunk 5]
```diff
@@ -0,0 +1,356 @@
+#   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserve.
```
Comment (wanghaoshuang) [inline paddleslim/core/layers.py:1]: What's the purpose of layers in this file?  Please add some description for this module. You can make a reference to [google python style](https://google.github.io/styleguide/pyguide.html#382-modules).

**Review by wanghaoshuang — CHANGES_REQUESTED** [review]

File: paddleslim/analysis/flops.py (hunk 1 above)
Comment (ceci3) [reply paddleslim/analysis/flops.py]: 静态图的卷积的flops也分成了conv op和elementwise op两个op计算

File: paddleslim/analysis/flops.py (hunk 3 above)
Comment (ceci3) [reply paddleslim/analysis/flops.py:120]: 算的是scale * (x-mean)里面的乘法

File: paddleslim/core/layers.py (hunk 5 above)
Comment (ceci3) [reply paddleslim/core/layers.py:1]: 一些ont-shot的超网络在训练过程中的时候会不断调整训练参数的channel之类的，这些感觉可以作为通用op，支持搜索用这些op

**Pull request closed without merging**